ownCloud/Settings/Passcode/PasscodeManager.swift

```diff
@@ -0,0 +1,375 @@
+        return (OCAppIdentity.shared().keychain.readDataFromKeychainItem(forAccount: passcodeKeychainAccount, path: passcodeKeychainPath) != nil)
+    }
+
+    private var passcodeFromKeychain: String? {
+        if let passcodeData = OCAppIdentity.shared().keychain.readDataFromKeychainItem(
+            forAccount: passcodeKeychainAccount, path: passcodeKeychainPath) {
+            return NSKeyedUnarchiver.unarchiveObject(with: passcodeData) as? String
+        } else {
+            return nil
+        }
+    }
+
+    private var shouldBeLocked: Bool {
+        var output: Bool = true
+
+        if isPasscodeActivated {
+            if let date = self.datePressedHomeButton {
+
+                let elapsedSeconds = Date().timeIntervalSince(date)
+                let minSecondsToAsk = self.userDefaults?.integer(forKey: SecuritySettingsFrequencyKey)
+
+                if Int(elapsedSeconds) < minSecondsToAsk! {
+                    output = false
+                }
+            }
+        } else {
+            output = false
+        }
+
+        return output
+    }
+
+    // MARK: - Init
+
+    static var sharedPasscodeManager = PasscodeManager()
+
+    public override init() {
+        // TODO: Use OCAppIdentity-provided user defaults in the future
+        self.userDefaults = UserDefaults(suiteName: OCAppIdentity.shared().appGroupIdentifier) ?? UserDefaults.standard
+
+        // Brute Force protection
+        self.timesPasscodeFailed = self.userDefaults!.integer(forKey: TimesPasscodeFailedKey)
+        if let data = self.userDefaults!.data(forKey: DateAllowTryPasscodeAgainKey) {
+            self.dateAllowTryAgain = NSKeyedUnarchiver.unarchiveObject(with: data) as? Date
+        }
+
+        super.init()
+    }
+
+    // MARK: - Show Passcode View
+
+    func showPasscodeIfNeeded(hiddenOverlay:Bool) {
+
+        if isPasscodeActivated {
+            if self.passcodeViewController == nil {
+
+                self.completionHandler = {
+                    self.datePressedHomeButton = nil
+                    self.timesPasscodeFailed = 0
+                    self.dismissPasscode()
+                }
+
+                self.passcodeViewController = PasscodeViewController(hiddenOverlay:hiddenOverlay)
+                self.createLockWindow()
+
+                // Brute force protection
+                if let date = self.dateAllowTryAgain, date > Date() {
+                    //User killed the app
+                    self.passcodeMode = .unlockPasscodeError
+                    self.passcodeViewController?.setEnableNumberButtons(isEnable: false)
+                    self.scheduledTimerToUpdateInterfaceTime()
+                } else {
+                    self.passcodeMode = .unlockPasscode
+                }
+
+                self.updateUI()
+
+            } else {
+                self.passcodeViewController?.showOverlay()
+            }
+        }
+    }
+
+    func showAddOrEditPasscode(completionHandler: @escaping CompletionHandler) {
```

Comment:
I think here fits better `showPasscode`

Reply:
We have two differents "showPasscode":
- showPasscodeIfNeeded: Passcode to be unlock
- showAddOrEditPasscode: Passcode to add or edit (delete)

I will change the name to showAddOrDeletePasscode because the true it is that you can not make a real "edit"